Feat  bind event listener to return key

changangus opened the pull request:
## Description
Add event listeners to both the return keypress and submit event. Added tests for these event listeners.

## Spec

See Issue: [FSA21V2-62](https://sparkbox.atlassian.net/browse/FSA21V2-62)

## Validation
<!-- delete anything irrelevant to this PR -->

- [x] This PR has code changes, and our linters still pass.
- [x] This PR has new code, so new tests were added or updated, and they pass.
- [ ] This PR affects production code, so it was browser tested (see below).

### To Validate

1. Make sure all PR Checks have passed.
1. Pull down all related branches.
1. Navigate to... _[continue instructions here]_

### Browser Testing
<!-- Delete if irrelevant to this issue -->

#### macOS

- [ ] Chrome, current release
- [ ] Firefox, current release
- [ ] Safari, current release

#### Windows

- [ ] Chrome, current release
- [ ] Firefox, current release
- [ ] Edge, current release

#### Mobile

- [ ] Mobile Safari on iPad
- [ ] Mobile Safari on iPhone

Commit:
feat: add event binders for 'Enter' & 'Submit'

The functions console log the text in the textarea element
and reset the value to be an empty string

FSA21V2-62

Commit:
refactor: change eventHandlers to accept callbacks

eventHandler functions now accept callbacks as a parameter

FSA21V2-62

Commit:
test: add missing eventHandler tests

added tests for both event handler functions

FSA21V2-62

Commit:
fix: index.spec.js restructure

Tests were failing in index.spec.js because there
was no document was defined. the code from index.js was
abstracted into a exported function called main in app.js.

FSA21V2-62

Commit:
refactor: change key event 'keyCode' to 'key

Instead of using the keyCode value for a keypress, we
use key for readability. Now instead of 'keyCode === 13',
it will read 'key === "Enter"'.

FSA21V2-62

robtarr commented on src/js/eventHandler.spec.js:
```diff
@@ -0,0 +1,45 @@
+</form>
+`;
+
+describe('eventHandler.js', () => {
```
```suggestion
describe('#callbackWhenEnterIsPressed', () => {
```

The entire file should be considered to be the tests for `eventHandler.js`, this shows that this block of tests apply to a specific function. I would wrap the test(s) for the other function in a similar `describe` block.

robtarr commented on src/js/eventHandler.spec.js:
```diff
@@ -0,0 +1,45 @@
+`;
+
+describe('eventHandler.js', () => {
+  it('Should run callback and set textarea to empty string on "Enter" key press', () => {
```
It would be nice to split this into 2 separate tests.  One for calling the `callback` function and one for resetting the text. That way each test is only concerned with one thing.

robtarr commented on src/js/eventHandler.spec.js:
```diff
@@ -0,0 +1,45 @@
+    expect(callback).toHaveBeenCalled();
+  });
+
+  it('Should run callback and set textarea to empty string on submit event', () => {
```
It would be nice to split this into 2 separate tests.  One for calling the `callback` function and one for resetting the text. That way each test is only concerned with one thing.

robtarr commented on src/js/eventHandlers.js:
```diff
@@ -0,0 +1,20 @@
+export const callbackWhenEnterIsPressed = (callback) => {
+  const textArea = document.querySelector('.chat-window__input');
```
I would recommend moving the setting of `textArea` outside of the function so that you don't have to query the DOM for it every time you need it.

robtarr replied:
This ended up causing problems with the jsdom mock DOM not being present when the selector ran, so we ignored it.

robtarr commented on src/js/index.js:
```diff
@@ -1,5 +1,10 @@
 
 const helloWorld = () => console.log('Hello!');
 
+callbackWhenEnterIsPressed(addMessageToChatHistory);
```
In the truest sense, these aren't callback functions. I think the real intent of these functions is to establish the event listeners.

changangus replied:
Yeah, I think that naming is a bit confusing. I renamed them to onEnterPress and onFormSubmit, Alice and I thought it was easier to read something like "onEnterPress(addMessageToChatHistory)".

robtarr commented on src/js/index.spec.js:
```diff
@@ -1,4 +1,7 @@
-import './index.js';
+/**
```
Do you need `jsdom` loaded in this placeholder test file?

jordanleven replied:
Ah, I see what you were talking about @changangus. This issue is because JDOM will mock out basic DOM elements, but not the `addEventListener` method on elements.

One approach you could take would be to do a small wrapper implementation around the `addEventListener` API that would allow you to mock out the contents of the wrapper (`jest.mock('./path/to/addEventListener/mock`) while testing the rest of the code. This, I think, would be the easiest way forward.

Another approach would be to _completely_ mock out the `eventHandlers.js` module with something like [mock module](https://jestjs.io/docs/mock-functions#mocking-modules) or [mock implementation](https://jestjs.io/docs/mock-functions#mock-implementations), but that would mean that tests on index.js wouldn't _actually_ be importing the real version of `eventHandlers`, which would open up the possibility of passing tests when the application code fails.

Lastly, another approach would include globally mocking out the `addEventListener` to mocked out elements in the DOM, but that would require mocking them out in _all_ tests that import the `eventHandlers` file which would get tedious and messy relatively quickly.

Commit:
refactor: eventHandler.spec.js

When testing each eventHandler we now check if the callback is
called in one test and that the textarea is reset in another.

FSA21V2-62

Commit:
refactor: eventHandler.js

changed the names of each function

FSA21V2-62

Commit:
test: test app.js

add tests to app.spec.js to make sure all actions
get run when we call main.

FSA21V2-62

dustin-jw commented on src/js/app.js:
```diff
@@ -0,0 +1,12 @@
+import { onEnterPress, onFormSubmit } from './eventHandlers';
+
+export default function main() {
+  const helloWorld = () => console.log('Hello!');
```
You can probably ditch the `helloWorld` stuff, since you've got an app that's doing real things now.

dustin-jw commented on src/js/app.spec.js:
```diff
@@ -0,0 +1,20 @@
+jest.mock('./eventHandlers.js');
+jest.mock('./chatWindow.js');
+
+describe('app.js', () => {
```
I'd maybe rename this to something more descriptive, like "app initialization" or something. The file naming convention already implies that this is for `app.js`.

dustin-jw commented on src/js/app.spec.js:
```diff
@@ -0,0 +1,20 @@
+jest.mock('./chatWindow.js');
+
+describe('app.js', () => {
+  it('Runs onEnterPress with addMessageToChatHistory as a parameter', () => {
+    main();
+    expect(onEnterPress).toHaveBeenCalledWith(addMessageToChatHistory);
+  });
+  it('Runs onFormSubmit with addMessageToChatHistory as a parameter', () => {
+    main();
+    expect(onFormSubmit).toHaveBeenCalledWith(addMessageToChatHistory);
+  });
```
This is just a code-style preference thing, but I'd add an empty line between the tests to break them up visually. Literally just nitpicking because everything else is so solid 😂

changangus replied:
I agree that it makes it easier to read, added this as well!

dustin-jw commented on src/js/eventHandler.spec.js:
```diff
@@ -0,0 +1,67 @@
+ * @jest-environment jsdom
+ */
+
+import { onFormSubmit, onEnterPress } from './eventHandlers';
```
Just for keeping the naming convention consistent, I'd probably rename this file to `eventHandlers.spec.js`.

changangus replied:
Whoops! I didn't notice it was missing the 's', just changed it!

dustin-jw reviewed (commented):
This looks super solid! All of my feedback is about minor code style and naming things – overall, this is really good stuff. 🏆

Commit:
refactor: cleanup app.js & app.spec.js

Got rid of hello world console log, created some more whitespace in
tests & renamed the description block to be more clear.

FSA21V2-62

dustin-jw reviewed (approved):
Looks good, awesome work! 🎉
